Twilio Lookup to Validate Phone Numbers

stevejgordon opened the pull request:
Fixes #1680

This code adds a service wrapper around the Twilio Rest API client. It uses the Lookup service to validate phone number formats and types.

It also ensures we save the correct E.164 formatted number to ensure our SMS response code can find matching requests.

Includes a faked version of the service for dev use - doesn't require Twilio account.

Currently a WIP for initial feedback and discussion. Needs to be implemented for CSV and manual request creation + unit testing and general checking through. Also needs some code to convert Country strings to ISO format country codes.

Tested against the live Twilio service using a real number and it worked as expected.

mgmccarthy commented:
@stevejgordon I'm going to pull this locally tonight or tomorrow and take a look. I'll get back to you. Thanks!

mgmccarthy commented:
@stevejgordon, the code looks solid, thanks for figuring all this Twilio stuff out. A couple questions:

1. do you need me to take care of updating the CSV request import and UI request creation with your Twilio functionality?
2. do you know if there is a request limit on this Twilio API (for example, Twilio only supports "x" amount of API invocations for a given IP address over a time period of "y") where we would need to worry about throttling the amount of api invocations for csv and api request imports?
3. Do you think that the Request model/entity should be updated to include more than one phone number field? For one Request, depending on the integration, we might be given both a landline and a mobile number for example. In the future, I know we want the request confirm process to be robust where we fallback to different communication types if one type "fails" or we might not have the data on the incoming request to try different types (for example, if we're not getting a landline, we can't fall back to a phone call via an IVR system)

wondering what your thoughts were.

Thanks!

stevejgordon commented:
Thanks @mgmccarthy for the review. On the basis of your thumbs up I'll look to get this rebased and a few tests added so that I can take this out of WIP.

Answers to your questions

1. That would be great. I'll leave that to you once this is finalised and merged in if that's okay. I can create an issue to cover that work.
2. I wasn't able to find any rate limits listed for the lookup API. The docs were a little slim, or I was looking in the wrong place.
3. It's possible we'll need a landline, but I guess until we have functionality around it / a new organization which needs it for integration it's a case of YAGNI. @tonysurma - Any requirements for this in the near future do you think?

tonysurma commented:
For two, you are good, we are charged per call so they are okay with all but crazy levels of calls. 

For three, stay as is and keep simple, new requirements can come later, if they do at all for that.

stevejgordon commented:
Thanks @tonysurma - I'll get this finalised for inclusion ASAP

stevejgordon commented:
Actually, I found this info:

> Twilio's Lookup API currently only handles a single request per phone number. If you have a batch of several phone numbers you would like to lookup, we recommend that you make multiple requests to Twilio's API. Please keep in mind that each Twilio account may only make 100 concurrent API requests per second.

So on CSV we need to be wary of it. Otherwise I think we're okay.

tonysurma commented:
@stevejgordon ah interesting, they have a slightly different plan for the lookup API then.  Agree only on CSV and then that is still pretty high so just need to watch for batches greater than 100 calls.  Thanks for researching.

stevejgordon commented:
@tonysurma no problem. The easiest option might be to wrap the code with Polly; a retry framework and then if we get any failed responses from the endpoint we can try again after a short delay. Covers us for short outages also.

mgmccarthy commented:
@tonysurma 

> or two, you are good, we are charged per call so they are okay with all but crazy levels of calls.

any example of "crazy level of calls" is something imports a CSV of request with tens of thousands of rows... we don't throttle nor offload asynchronously any of this work, it's API invocation for each request in a for loop.

stevejgordon commented:
@mgmccarthy I've refactored slightly for testing and added the tests. I think this is good to go for now. Can open a new issue to consider CSV import rate limiting.

@tonysurma If Mike's happy then this is ready to merge. Removed the WIP title.

mgmccarthy commented:
@stevejgordon @tonysurma I'll see if there is a way to "throttle" the processing of CSV imports for Requests when I plug the `ValidatePhoneNumberRequest` into the CSV code.

stevejgordon commented:
@mgmccarthy Was thinking the same thing!

tonysurma commented:
@stevejgordon @mgmccarthy - good to merge?

mgmccarthy commented:
@tonysurma yep
